src/toposort.rs

```diff
@@ -193,60 +196,73 @@ impl TopologicalSorter {
     ///     of the nodes given to :meth:`done`.  This can only happen if the ``initial`` nodes had
     ///     even a partial topological ordering amongst themselves, which is not a valid
     ///     starting input.
-    fn done(&mut self, py: Python, nodes: Vec<usize>) -> PyResult<()> {
-        let dag = &self.dag.borrow(py);
-        let (in_dir, out_dir) = traversal_directions(self.reverse);
-        for node in nodes {
-            let node = NodeIndex::new(node);
-            match self.node2state.get_mut(&node) {
-                None => {
-                    return Err(PyValueError::new_err(format!(
-                        "node {} was not passed out (still not ready).",
-                        node.index()
-                    )));
-                }
-                Some(NodeState::Done) => {
-                    return Err(PyValueError::new_err(format!(
-                        "node {} was already marked done.",
-                        node.index()
-                    )));
-                }
-                Some(state) => {
-                    debug_assert_eq!(*state, NodeState::Ready);
-                    *state = NodeState::Done;
-                }
+    fn done(&mut self, nodes: &Bound<PyAny>) -> PyResult<()> {
+        if let Ok(node) = nodes.extract::<usize>() {
+            self.done_single(nodes.py(), NodeIndex::new(node))
+        } else if let Ok(nodes) = nodes.downcast::<PyList>() {
+            for node in nodes {
+                self.done_single(nodes.py(), NodeIndex::new(node.extract()?))?
+            }
+            Ok(())
+        } else {
+            for node in nodes.iter()? {
+                self.done_single(nodes.py(), NodeIndex::new(node?.extract()?))?
+            }
+            Ok(())
+        }
```

Comment:
Not quite sure what's the difference in the 2 else clauses. Is the last one assuming `nodes` can't be downcast to `PyList`? If so, what's an example?

Reply:
The Rust-space typecast is stricter than a Python-space call to `list` (which copies), so anything iterable that's not a list or a subclass would fail it, like a tuple for example.